code/modules/transport/pods/MainWeapon.dm

```diff
@@ -380,6 +380,624 @@
+					dat+="<B>Repair</B><BR>"
+					dat+="<A href='?src=\ref[src];cutter=1'>Switch to Drilling</A><BR>"
+				else
+					dat+="<A href='?src=\ref[src];floors=1'>Switch to Floors</A><BR>"
+					dat+="<A href='?src=\ref[src];r_floors=1'>Switch to Reinforced Floors</A><BR>"
+					dat+="<A href='?src=\ref[src];walls=1'>Switch to Walls</A><BR>"
+					dat+="<A href='?src=\ref[src];repair=1'>Switch to Repair</A><BR>"
+					dat+="<B>Drilling</B><BR>"
+			dat+="<BR>"
+			dat+="Wide Field Mode <B>[src.wide_field ? "Active" : "Inactive"]</B> <A href='?src=\ref[src];wide_field=1'>(Toggle)</A><BR>"
+			dat+="[src.steel_sheets.amount] of [src.max_sheets] Steel Sheets Loaded <A href='?src=\ref[src];load=1'>(Load)</A><BR>"
+
+		else
+			dat += {"<B><span style=\"color:red\">SYSTEM OFFLINE</span></B>"}
+		user.Browse(dat, "window=ship_main_weapon")
+		onclose(user, "ship_main_weapon")
+		return
+
+	Topic(href, href_list)
+		if(..())
+			return
+
+		if (href_list["floors"])
+			mode = EFIF_MODE_FLOORS
+			firerate = 15
+
+		else if (href_list["r_floors"])
+			mode = EFIF_MODE_R_FLOORS
+			firerate = 25
+
+		else if (href_list["walls"])
+			mode = EFIF_MODE_WALLS
+			firerate = 35
+
+		else if (href_list["repair"])
+			mode = EFIF_MODE_REPAIR
+			firerate = 15
+
+		else if(href_list["cutter"])
+			mode = EFIF_MODE_DRILL
+			firerate = 12
+
+		else if(href_list["wide_field"])
+			wide_field = !wide_field
+
+		else if(href_list["load"])
+			src.scoop_up_sheets()
+
+		opencomputer(usr)
+		return
+
+/obj/overlay/construction_field
+	name = "energy"
+	icon = 'icons/obj/objects.dmi'
+	icon_state = "buildeffect"
+	layer = EFFECTS_LAYER_BASE
+	///Tracks the build type of what each field is building for cost calculation and placement logic
+	var/to_build
+	///If in repair mode, build the precise initial turf
+	var/repair_mode = FALSE
+
+/datum/action/bar/construction_field
+	var/obj/machinery/vehicle/pod
+	var/obj/item/shipcomponent/mainweapon/constructor/buildtool
+	var/mode
+	///Sum of current attempted construction's cost in sheets
+	var/action_build_cost = 0
+	duration = 1 SECONDS
+
+	New(var/obj/thepod,var/obj/thetool,var/passmode)
+		src.pod = thepod
+		src.buildtool = thetool
+		src.mode = passmode
+		for (var/obj/overlay/construction_field/F in buildtool.active_fields)
+			src.duration += src.get_duration_mod(F)
+			src.action_build_cost += buildtool.get_cost_mod(F)
+		..()
+
+	proc/get_duration_mod(var/obj/overlay/construction_field/F)
+		switch(F.to_build)
+			if(EFIF_MODE_FLOORS)
+				return EFIF_FLOOR_BUILD_TIME
+			if(EFIF_MODE_R_FLOORS)
+				return EFIF_R_FLOOR_BUILD_TIME
+			if(EFIF_MODE_WALLS)
+				var/space_adjustment = EFIF_WALL_BUILD_TIME
+				if(istype(get_turf(F),/turf/space))
+					space_adjustment += EFIF_FLOOR_BUILD_TIME
+				return space_adjustment
+
+	///Check whether dense objects are present in build fields; interrupt any fields which have been blocked
+	proc/dense_object_refresh()
```

Comment:
Maybe this check should only happen for walls? It's a bit odd not to be able to repair floors because there's an unexploded pod/fuel tank over where they should be.

Reply:
Good thought. Will take a poke at adjusting that.

Reply:
Adjusted.